pkg/util/mount/types/types.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
```

Comment:
I really don't like this pattern of stuffing all interfaces in to a single package.

A much cleaner pattern is:

- github.com/bar/foo
  - foo.go
    - `type Foo interface` - the interface that all external references should depend on
    - `func NewFoo(...) Foo` - returns a concrete implementation of interface Foo
  - foo_test.go
    - Tests the concrete implementation of Foo
  - fake_foo.go
    - `func NewFakeFoo(...) Foo` - returns a fake implementation of interface Foo

This makes it dead simple to use, if you want foo (real or a fake), import "github.com/bar/foo".

Looking at the comment that spurred this refactor (https://github.com/kubernetes/utils/pull/100#discussion_r311318715), the problem there was that the package k8s.io/utils/exec doesn't implement a fake, so a thin wrapper `mounter/exec.go` was implemented. The fix for that is not to stuff all the types in to a single file. We should look in to if we really need the wrapper "mount/exec" Interface around "k8s.io/utils/exec"? Or can we use "k8s.io/utils/exec" directly? You could create a package just for the exec fake that implements a fake for the interface in utils/exec without creating a wrapper for the non-test code OR even better, put the fake in the utils/exec package.

Reply:
Also I realize this is a big change, after all the heroic work you've done. So please feel free to push back. My thinking is if we're gonna refactor this, let's refactor it right.

Reply:
I hear you. And FWIW, I agree with your package structure. Either direction we go is a big change. I just want to make sure we have an agreed upon approach that will also satisfy @thockin desires before putting in more effort.

There are also "fake" implementations for the mounter and the HostUtils interface. Whether those are in a `testing` subpackage or not is not super important to me (but like I said, I have a slight preference for your approach, Saad), I just want to make sure that whatever we produce will be approved at a higher level.

I'm in full agreement with "let's refactor it right". This is a good opportunity to clean up a lot of things. My only concern is getting it in for 1.16. I know a lot of interested parties are hoping for that. The amount of code these changes touch is massive, which leads to difficulty getting `/approve`'s. There are definitely tradeoffs. But as long as it's possible, let's do it right.

Reply:
The ugliness that our patch depends on, vendoring utils/mount in staging/openstack, is by design a short-lived hack and will be removed almost immediately after 1.16 release. While this refactor would be a "nice-to-have" for us, I don't think it's a "must-have".

Reply:
@saad-ali Any agreement between you and Tim on preferred package structure here?

Reply:
I was able to sync up with @thockin. He pointed me to the following go lang principles.
1) "Accept interfaces, return concrete types" (see [1](http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/) and [2](https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8)).
2) "Consumer should define the interface" (see [1](https://github.com/golang/go/wiki/CodeReviewComments#interfaces) and [2](https://twitter.com/davecheney/status/942593128355192832?s=20)).

"Accept interfaces" makes perfect sense.

"Return concrete types" and "consumer should define the interface" were new to me. The argument for doing so is that "defining an interface upfront is usually a code smell for overengineering" and it not doing so "reduces dependencies" (see [1](https://blog.chewxy.com/2018/03/18/golang-interfaces/) and [2](https://news.ycombinator.com/item?id=16623720)).

In this case, given that we have multiple concrete implementations of the mounter: real (linux, windows) and fake. And we our external consumers will want to use both our real and fake implementations, we have sufficient justification to define an interface for Mounter and export fake and real concrete implementations.

Therefore, as far as the file structure is concerned, for the top level mounter package, we should do what I recommended above ("Proposal 1") with a small modification, the return type for `NewFoo(...)` and `NewFakeFoo(...)` should be the concrete type not an interface.

For sub-packages that mounter will depend on, you can follow the same pattern as long as there are multiple consumers of both real and fake types. If not, then we should follow go lang guidance.

Reply:
Thanks for the clarification, and the links for reference.

I will refactor this PR after https://github.com/kubernetes/kubernetes/pull/81794 goes in, as that will also reduce the scope of what's happening. In fact, I already have a follow-on commit to that one that does most of what we need for `HostUtils`, but not yet for `Mount`.

Progress!

Reply:
Sounds good, thanks a lot for your patience and hard work!!